src/main/java/com/google/codeu/data/Datastore.java

```diff
@@ -85,6 +87,59 @@ public void storeMessage(Message message) {
     return messages;
   }
 
+  /**
+   * Gets messages sent between the logged-in user and another user.
+   *
+   * @return a list of messages sent between the logged-in user and another user, or empty list
+   * if logged-in user or other user have never sent a message to each other. List is sorted by time ascending.
+   */
+  public List<Message> getMessagesBetweenTwoUsers(String loggedInUser, String otherUser) {
+    List<Message> messages = new ArrayList<>();
+
+    // Messages between two people, where logged-in user is the sender
+    Filter messagesSentByLoggedInUser = new Query.FilterPredicate("user", FilterOperator.EQUAL, loggedInUser);
+    Filter messagesReceivedByOtherUser = new Query.FilterPredicate("recipient", FilterOperator.EQUAL, otherUser);
+
+    // All the messages sent by logged-in user to the other user
+    Filter loggedInUserMessages = CompositeFilterOperator.and(messagesSentByLoggedInUser, messagesReceivedByOtherUser);
+
+    // Messages between two people, where logged-in user is the recipient
+    Filter messagesSentByOtherUser = new Query.FilterPredicate("user", FilterOperator.EQUAL, otherUser);
+    Filter messagesReceivedByLoggedInUser = new Query.FilterPredicate("recipient", FilterOperator.EQUAL, loggedInUser);
+    
+    // All the messages recieved by logged-in user sent by the other user
+    Filter otherUserMessages = CompositeFilterOperator.and(messagesSentByOtherUser, messagesReceivedByLoggedInUser);
+
+    // All the messages between the two users
+    Filter directMessages = CompositeFilterOperator.or(loggedInUserMessages, otherUserMessages);
+
+    Query query =
+        new Query("Message")
+            .setFilter(directMessages)
+            .addSort("timestamp", SortDirection.ASCENDING);
+    PreparedQuery results = datastore.prepare(query);
+
+    for (Entity entity : results.asIterable()) {
+      try {
+        String idString = entity.getKey().getName();
+        UUID id = UUID.fromString(idString);
+        String text = (String) entity.getProperty("text");
+        long timestamp = (long) entity.getProperty("timestamp");
+        String sender = (String) entity.getProperty("user");
+        String receiver = (String) entity.getProperty("receiver");
+        Message message = new Message(id, sender, text, timestamp, receiver);
+
+        messages.add(message);
+      } catch (Exception e) {
```

Comment:
This kind of general catch statement for debug logging is OK at higher-levels in the stack (e.g. if you want to create a `ServletBase` class that all the servlets in this application extend and that injects this kind of generic error-handling for uncaught exceptions), but is a pattern that is best avoided otherwise. Other than in framework code high up in the stack, functions should generally catch narrowly-defined exception type(s).

Reply:
I'm not sure I understand this

Reply:
What would be a better catch statement?

Reply:
If you want to do this kind of general catch for debugging, I would simply move it higher-up in the stack; that is, there is a function called `service`:

https://tomcat.apache.org/tomcat-5.5-doc/servletapi/javax/servlet/http/HttpServlet.html#service(javax.servlet.http.HttpServletRequest,%20javax.servlet.http.HttpServletResponse)

... that is responsible for calling `doGet`, `doPost`, etc. 

I would do something like defining:

     com.google.appengine.api.utils.SystemProperty;

     class CodingWithClassServletBase extends HttpServlet {
         @Override
          protected void service(HttpServletRequest req, HttpServletResponse resp)
                throws ServletException, java.io.IOException {
             try {
               super.service(req, resp);
             } catch (Exception e) {
               if (isDebugMode()) {
                  // When running locally, apply special error handling to make it
                  // easier to see the exception that was thrown.
                   e.printStackTrace();
                   resp.setStatus(500);
               } else {
                   // In the production environment, use the standard built-in error handling.
                   throw e;
               }
          }

          private boolean isDebugMode() {
            return SystemProperty.environment.value() != SystemProperty.Environment.Value.Production;
          }
      }

... and then I would have this class extend `CodingWithClassServletBase` instead of `HttpServlet`.

If you have exception handling that is specific to `getMessagesBetweenTwoUsers`, then you should catch specific exception types (e.g. `ClassCastException`, `IOException`, etc. as appropriate) and not the `Exception` base class.

Reply:
Alright, I think I was able to implement the changes you suggested. Let me know what you think